subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/CrossTaskGroovyJavaJointIncrementalCompilationIntegrationTest.groovy

```diff
@@ -0,0 +1,118 @@
+        given:
+        // A is a private dependency of B and B is referenced in E.isCacheEnabled through inheritance
+        File aClass = sourceForLanguageForProject(CompiledLanguage.JAVA, "api", "class A { void m1() {}; }")
+        sourceWithFileSuffixForProject(bSuffix, "impl", "$bCompileStatic class B { void m1() { A a = new A(); a.m1(); }; }")
+        sourceWithFileSuffixForProject("java", "impl", "class C extends B {}")
+        sourceWithFileSuffixForProject("java", "impl", "class D extends C { static boolean getCache() { return true; } }")
+        File eClass = sourceWithFileSuffixForProject("groovy", "impl", "class E { boolean isCacheEnabled = D.cache }")
+        run ":impl:compileGroovy"
+
+        when:
+        aClass.text = "class A { void m1() {}; void m2() {}; }"
+        eClass.text = "class E { boolean isCacheEnabled = D.cache; int a = 0; }"
+        impl.snapshot()
+
+        then:
+        run ":impl:compileGroovy"
+        impl.recompiledClasses(*expectedRecompiledClass)
+
+        where:
+        bSuffix  | bCompileStatic                    | expectedRecompiledClass
+        "java"   | ""                                | ["B", "C", "D", "E"]
+        "groovy" | ""                                | ["B", "E"]
+        "groovy" | "@groovy.transform.CompileStatic" | ["B", "E"]
+    }
+
+    def 'inc. compilation does not fail on api change when we compile only groovy and affected class is #bCompileStatic #bSuffix'() {
+        given:
+        buildFile << """
+            allprojects {
+                tasks.withType(GroovyCompile) {
+                    options.incremental = true
+                    groovyOptions.fileExtensions = ['groovy']
+                }
+            }
+        """
+        // A is a private dependency of B and B is referenced in E.isCacheEnabled through inheritance
+        File aClass = sourceForLanguageForProject(CompiledLanguage.JAVA, "api", "class A { void m1() {}; }")
+        sourceWithFileSuffixForProject(bSuffix, "impl", "$bCompileStatic class B { void m1() { A a = new A(); a.m1(); }; }")
+        sourceWithFileSuffixForProject("java", "impl", "class C extends B {}")
+        sourceWithFileSuffixForProject("java", "impl", "class D extends C { static boolean getCache() { return true; } }")
+        File eClass = sourceWithFileSuffixForProject("groovy", "impl", "class E { boolean isCacheEnabled = D.cache }")
+        run ":impl:compileGroovy"
+
+        when:
+        aClass.text = "class A { void m1() {}; void m2() {}; }"
+        eClass.text = "class E { boolean isCacheEnabled = D.cache; int a = 0; }"
+        impl.snapshot()
+
+        then:
+        run ":impl:compileGroovy"
+        impl.recompiledClasses(*expectedRecompiledClass)
+
+        where:
+        bSuffix  | bCompileStatic                    | expectedRecompiledClass
+        "java"   | ""                                | ["E"]
```

Comment:
Why isn't `B` recompiled here when it is a Java class?

Reply:
We recompile A since it's in a different project and it compiled with `compileJava`. But we don't compile `B.java`, since only `.groovy` are recompiled. So classes D, C and B are never compiled and Groovy for some reason doesn't complain about it at compile time.

Reply:
So that test basically shows that if we don't have a class/source for something at all, then dynamic Groovy still compiles? So you could do the same when only having class `E` with the same contents and none of the other classes in the `impl` subproject?

Reply:
It looks like yes

Reply:
I guess this is true for calls/usages of some class, that Groovy compiles to dynamic code.

Reply:
I think we should simplify the test in that case and remove all the Java files from it.